ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,647 @@
+MSMC or the SMC with multiplicity constraints problem".
+
+It /may/ be the case that our problem is particularly degenerate, and so easier
+to solve. In particular, our variables are binary, their coefficients in the
+constraints are binary, and in fact our constraints form a banded matrix. Only
+the lower bound of each constraint can be greater than 1. But I have not yet
+recognized how to do it.
+
+The closest I have come to a solution is an incremental algorithm that toggles
+one empty slot at a time. I /think/ the completeness of this algorithm requires
+that the next toggle is one of the empty slots that occurs in the most number
+of sparse windows. However, I have not proved that. Moreover, even if it is
+true, a counterexample proves that choosing an arbitrary such slot is not
+sufficient:
+
+>     A   B   C
+> -----   -----
+> 01110111011101110
+>     -----   -----
+
+In this example, only slots A, B, and C occur in two sparse windows, so those
+are the candidates for the next toggle by my conjecture. If we suppose that
+each window is only missing one active slot (eg need 4 actives per 5 slots),
+then toggling B would require three toggles in total, whereas toggling A and C
+would solve the problem with just two toggles.
+
+-}
+
+-- | A 'ChainSchedule' that satisfies 'checkHonestChain'
+--
+-- The distribution this function samples from is not simple to describe. It
+-- begins by drawing a sample of length 'Len' from the Bernoulli process
+-- induced by the active slot coefficient 'Asc'. (IE 'Len' many i.i.d. samples
+-- from @Uniform(f)@). Then it slides an @s-d@ window along that leader
+-- schedule, toggling as many empty slots as necessary to ensure every such
+-- window has at least @k+1@ active slots, ie Enriched Praos Chain
+-- Growth---HOWEVER, it is a difficult computation to add a /strictly minimal/
+-- number of active slots to the output of the Bernoulli process (it's an
+-- integer programming optimization problem as far as we can tell). We have
+-- settled for a relatively simple algorithm that approaches the minimal number
+-- of toggles (TODO calculate how tight the bounds are).
+--
+-- NOTE: when @'Asc' >> (k+1)/(s-d)@ the sample is most likely directly from
+-- the Bernoulli process.
+--
+-- NOTE: when @'Asc' = 0@ the resulting leader schedule is periodic and has
+-- exactly @k+1@ active slots in every @s-d@ window.
+--
+-- NOTE: when @'Asc' << (k+1)/(s-d)@ the sample is likely to be the
+-- concatentation of a few long periodic and minimal runs with the occasional
+-- window with @k+2@ active slots demarcating the runs.
+--
+-- TODO sample from a disjunction of generators that explore interesting
+-- boundary conditions. EG one would fill in each sparse window by setting the
+-- /youngest/ empty slots in it.
+uniformTheHonestChain ::
+  forall base hon g.
+     R.RandomGen g
+  => Maybe Asc   -- ^ 'Nothing' means @0@, which induces a periodic chain
```

Comment:
Is it meaningful to have `Asc == 0`? I would imagine in such case no slot would be active.

Moreover, are `Nothing` and `Just 0` two representation of the same value in the domain?

Reply:
`Asc` cannot be `0`, by construction. (See its Haddock.)

Reply:
Ah, then the documentation of the `Maybe Asc` parameter makes no sense. `Nothing` can't mean `0`.

Reply:
Why can't `Nothing` mean `0`? The intention of feeding `Nothing` is to generate a chain where no block can be removed without violating any of the safety properties.

Reply:
`Nothing` cannot meant `Asc==0` because the meaning of `Asc` is not defined when it is `0`.  

> The intention of feeding Nothing is to generate a chain where no block can be removed without violating any of the safety properties.

But I think this "intention" is an alright meaning for `Nothing`. It just needs to be documented :)

Reply:
Adopted in 37b643d9.